zetaclient/evm_client.go

```diff
@@ -577,28 +570,33 @@ func (ob *EVMChainClient) observeOutTx() {
 			//FIXME: remove this timeout here to ensure that all trackers are queried
 			outTimeout := time.After(time.Duration(timeoutNonce) * time.Second)
 		TRACKERLOOP:
-			// Skip old gabbage trackers as we spent too much time on querying them
 			for _, tracker := range trackers {
 				nonceInt := tracker.Nonce
-				if nonceInt < lowestOutTxNonceToObserve[ob.chain.ChainId] {
-					continue
-				}
 				if ob.isTxConfirmed(nonceInt) { // Go to next tracker if this one already has a confirmed tx
 					continue
 				}
+				txCount := 0
+				var receipt *ethtypes.Receipt
+				var transaction *ethtypes.Transaction
 				for _, txHash := range tracker.HashList {
 					select {
 					case <-outTimeout:
-						ob.logger.ObserveOutTx.Warn().Msgf("observeOutTx timeout on chain %d nonce %d", ob.chain.ChainId, nonceInt)
+						ob.logger.ObserveOutTx.Warn().Msgf("observeOutTx: timeout on chain %d nonce %d", ob.chain.ChainId, nonceInt)
 						break TRACKERLOOP
 					default:
-						if ob.confirmTxByHash(txHash.TxHash, nonceInt) {
-							ob.logger.ObserveOutTx.Info().Msgf("observeOutTx confirmed outTx %s for chain %d nonce %d", txHash.TxHash, ob.chain.ChainId, nonceInt)
-							break
+						if recpt, tx, ok := ob.checkConfirmedTx(txHash.TxHash, nonceInt); ok {
+							txCount++
+							receipt = recpt
+							transaction = tx
+							ob.logger.ObserveOutTx.Info().Msgf("observeOutTx: confirmed outTx %s for chain %d nonce %d", txHash.TxHash, ob.chain.ChainId, nonceInt)
 						}
-						ob.logger.ObserveOutTx.Debug().Msgf("observeOutTx outTx %s for chain %d nonce %d not confirmed yet", txHash.TxHash, ob.chain.ChainId, nonceInt)
 					}
 				}
+				if txCount == 1 { // should be only one txHash confirmed for each nonce.
+					ob.SetTxNReceipt(nonceInt, receipt, transaction)
+				} else if txCount > 1 { // should not happen. We can't tell which txHash is true. It might happen (e.g. glitchy/hacked endpoint)
```

Comment:
Can't we create a single function for this logic to combine evm and bitcoin code?

Reply:
I would like to unify it initially. The logic for EVM chain is intended for `outTx confirmation` while the BTC logic is intended for `outTx inclusion`. For BTC chain, zetaclients usually interact with a pending block (the one being mined) to move the pending nonce forward. Also, use separate functions make sense because evm chain is different from btc chain, for example
1. EVM chain has native concept nonce and receipt while bitcoin hasn't
2. EVM chain usually has quick finality while bitcoin takes 10 minute to generate a new block.